C++/include/algorithm/sorting/bucket_sort.hpp

```diff
@@ -0,0 +1,111 @@
+/*
+    Bucket / Bin sort
+    ----------------------
+    Bucket sort, or bin sort, is an integer sorting algorithm that works by
+    distributing the elements of an array into a number of buckets. Each 
+    is then sorted individually using a different sorting algorithm.
+
+    This is a linear sorting algorithm for evenly distributed items.
+
+    Time complexity
+    ---------------
+    O(N ^ 2) worst case
+    O(N + N^2 / k + k) average case when k is the number of buckets, 
+    which is O(n) if K ~ N
+
+    Space complexity
+    ----------------
+    O(N)
+*/
+
+#ifndef BUCKET_SORT_HPP
+#define BUCKET_SORT_HPP
+
+#include <vector>
+#include <list>
+#include "utils.hpp"
+#include <algorithm>
+
+using std::vector;
+using std::list;
+
+using SortingFunction = void(*)(vector<int>&, int, bool);
+
+template<SortingFunction sort>
+void bucket_sort(vector<int>& values, const int order = 1, const bool to_show_state = false) {
+
+    if(values.empty()) {
+        return;
+    }
+
+    int max_value = values[0];
+    int min_value = values[0];
+    int negative_count = 0, positive_count = 0;
+
+    for(int i : values) {
+        max_value = std::max(max_value, i);
+        min_value = std::min(min_value, i);
+        negative_count += i < 0;
+        positive_count += i >= 0;
+    }
```

Comment:
Did you consider `std::min_element`, `std::max_element` and `std::count_if`?
(If this is intentional in order to iterate once, `std::for_each`?)

They say the standard library's loops are the best.

Reply:
I started with those, but couldn't go over this 4 times just for using algorithms. Actually I need to remove the <algorithms> include...